site/content/blog/2020-03-09-zap-ssrf-setup/index.md

```diff
@@ -0,0 +1,59 @@
+
+## Solution
+### TLDR:
+1. Have a domain like something.mydomain.org pointing to a server you own (further details below).
+1. Set ZAP to always use the same callback port and define the Remote Address as your domain something.mydomain.org. (Via the [Callback Options](https://www.zaproxy.org/docs/desktop/ui/dialogs/options/callback/) screen.) {{< img "images/callback-options-setup.png" >}}
+1. Forward requests received on the server to the local ZAP with SSH port forwarding:
+    ```sh
+    ssh -N -R 38193:localhost:38193 myUser@something.mydomain.org
+    ```
+
+### My Setup:
+
+For this setup you will need a DNS name and a server with a public address. If you don’t have a DNS name and you don’t want to spend money you can create one for free on https://duckdns.org (or numerous other free services). The options I use on ZAP are the ones on the TLDR image.
+
+1.  I have a domain like something.mydomain.org pointing to a [Google Cloud](https://www.freecodecamp.org/news/how-to-create-and-connect-to-google-cloud-virtual-machine-with-ssh-81a68b8f74dd/) server I own, this allows me to have a public IP.
+
+1. I set ZAP to always use the same callback port, for instance 38193. This is important because if the port for ZAP callbacks changes you need to check it every time you start ZAP and change the system command for the port selected. (Via the [Callback Options](https://www.zaproxy.org/docs/desktop/ui/dialogs/options/callback/) screen.)
+
+1. I setup the Remote Address to my domain something.mydomain.org. This will make ZAP scan rules create SSRF request URLs with the domain something.mydomain.org and the static port we previously defined. For instance the XXE scan rule will send payloads with external entities leveraging a URL such as: http://something.mydomain.org:38193/callback/XXESD215F. 
+
+1. If a vulnerability exists the server will make a request to http://something.mydomain.org:38193/callback/XXESD215F. This request will be made to my cloud server because it is the location my DNS is pointing to. To receive this request on ZAP you need to somehow redirect the request to your local machine. One way that doesn’t require a public address on your machine is to use [SSH port forwarding](https://www.ssh.com/ssh/tunneling/example). What it does is to connect to your server through SSH, listen to a port on the server and send all the traffic received on that port to a port on your machine. The command I use is: 
+    ```sh
+    ssh -N -R 38193:localhost:38193 myUser@something.mydomain.org
+    ```
+
+If you want to have a valid HTTPS certificate you can run some reverse proxy on your server (as Nginx) with a certificate, proxy the requests to other port, for instance 9999, and change the port forwarding to that address:
+```sh
+ssh -N -R 9999:localhost:38193 myUser@something.mydomain.org
+```
+
+__Troubleshot__:
```

Comment:
Make it a heading instead if bold?
Two Os in shoot 👍

Reply:
like ###Troubleshot ?

Reply:
`### Troubleshoot` I think